aclk/agent_cloud_link.c

```diff
@@ -327,59 +403,295 @@ struct aclk_query *aclk_queue_pop()
 // Need to check if additional logic should be added to make sure that there
 // is enough information to determine the base topic at init time
 
-// TODO: Locking may be needed, depends on the calculation of the base topic and also if we need to switch
-// that on the fly
 
-char *get_publish_base_topic(PUBLISH_TOPIC_ACTION action)
+char *create_publish_base_topic()
 {
-    static char *topic = NULL;
-
     if (unlikely(!is_agent_claimed()))
         return NULL;
 
     ACLK_LOCK;
 
-    if (unlikely(action == PUBLICH_TOPIC_FREE)) {
-        if (likely(topic)) {
-            freez(topic);
-            topic = NULL;
-        }
+    if (unlikely(!global_base_topic)) {
+        char tmp_topic[ACLK_MAX_TOPIC + 1], *tmp;
+
+        snprintf(tmp_topic, ACLK_MAX_TOPIC, ACLK_TOPIC_STRUCTURE, is_agent_claimed());
+        tmp = strchr(tmp_topic, '\n');
```

Comment:
I think this is too little too late. We should check it after loading and maybe in separate function validate_claiming_id? Additionally we should check for special characters like #/ etc.etc.

Reply:
therefore the function that is loading it should either check it by itself or call `validate_claiming_id` which then should check it more carefully (not just look for endline) and set agent state to unclaimed if anything suspicious found

Reply:
True this needs to go all the way back to the claiming code. I will open a new issue to handle it in the claim.c. Once we reach this point the agent is assumed claimed which should not happen of the id is anyway invalid

Reply:
Well spotted @underhood - yes there is a new requirement on the claiming id that it cannot use any character that would be invalid as a component in an MQTT topic. We need to validate when it is loaded in - but we should also reject it with an error at the point of running the the claim script.